### ISSUE TITLE
TypeScript is unable to tell `as const` type when `"isolatedDeclarations": true` and `satisfies` is used

### ISSUE THREAD

**Issue by Meligy** [microsoft/TypeScript#62800]
### 🔎 Search Terms

I'm trying to revive https://github.com/microsoft/TypeScript/issues/58397

Please be patient with me. I do realise it's a locked issue. I'll make my case. Promise.

### 🕗 Version & Regression Information

This is happening in 5.9.3, but it was reported earlier in 5.5.

Not sure when it started.

### ⏯ Playground Link

[One example](https://www.typescriptlang.org/play/?isolatedDeclarations=true#code/C4TwDgpgBAwg9gGzgJwPJmASzgOygXigHIAlCAEyKgB9iBxZCCHK2ogIQQFcJXjOAhgGMA1nyIB1ABaZgvANwAoZRAAeYFMChDcAZy0AzOHAJQA3oqhWoAehtQAKshBRgJgOYQtmHHIQJMXWYgqCkIRktrAQAuYiJlBMSkxQBfKAFdbT1gJUUjE10BLF0DTAhMsh1kcgAefWQfdwAaWEQUdCxcAD55IA)

But please see the other examples in this issue also.

### 💻 Code

The following code forces you to add a type when `as const` is meant to avoid that.

```ts
const xWorks =  { test: 'value' } as const;
const xDoesNotWork = { test: 'value' } as const satisfies { test: string };
```

Error:

> Variable must have an explicit type annotation with --isolatedDeclarations.(9010)

[Playground](https://www.typescriptlang.org/play/?isolatedDeclarations=true#code/MYewdgzgLgBAHgdRAJwNYRgXhjA3jKAU2gC4YByANwEMAbAV0PJgF8ZqNRIoBuAKC7R4AERDEAciChI0WPAWJQyVOo2ZsOMQbAjUoASwgAzfcXlFSMaMn1gA5ix5A)

There's a suggested workaround:

```ts
const xWorks =  { test: 'value' } as const;
const xDoesNotWork = { test: 'value' } as const;
xDoesNotWork satisfies { test: string };
```

But this loses one of the key reasons to use `satisfies`. You do not get intellisense. Let me show:

### 🙁 Actual behavior

Take the following example:

```ts
type ColorOption = 'Red' | 'Green' | 'Blue' | 'Black' | 'White';

export const foo = {
    // Try to get intellisense here
    a: ''
} as const;

foo satisfies Record<string, ColorOption>;
```

If you try to get intellisense inside the value of `a`, you get nothing.

<img width="1140" height="758" alt="Image" src="https://github.com/user-attachments/assets/bacb1589-d4b0-4725-a2fa-e45ef29e1f49" />

[Playground](https://www.typescriptlang.org/play/?isolatedDeclarations=true#code/C4TwDgpgBAwg9gGzgJwPJmASzgOygXigHIAlCAEyKgB9iBxZCCHK2ogIQQFcJXjOAhgGMA1nyIB1ABaZgvANwAoRRAAeYFMChDcAZy0AzOHAJQA3oqhWoAehtQAKshBRgJgOYQtmHHIQJMXWYgqCkIRktrAQAuYiJFAF8oAV1tPWAlRSMTXQEsXQNMCFSyHWRyAB59ZB93ABpYRBR0LFwAPnkgA)

This is much worse if the value is an object.

Like this example:

```ts
interface ListItem {
  name: string;
  description: string;
}

export const foo = {
  a: {
    // try to type `name` here (without ``):
  }
} as const;

foo satisfies Record<string, ListItem>;
```

It actively rejects your value:

<img width="1138" height="604" alt="Image" src="https://github.com/user-attachments/assets/1460953e-0fe6-470e-852e-2508b6db12e6" />

It does not get better with narrowing:

<img width="1116" height="718" alt="Image" src="https://github.com/user-attachments/assets/86040cd4-20f7-489c-ba39-8f4beeb7b9bb" />

Although once you are done, it'll recognise it.

<img width="788" height="632" alt="Image" src="https://github.com/user-attachments/assets/8dc8ca18-ba0b-4bc9-b3ff-509f47edb292" />

But you'll still have the same issue whenever you add other items:

<img width="1122" height="752" alt="Image" src="https://github.com/user-attachments/assets/51cd3890-7c0d-42e3-85c7-1d586513814a" />

<img width="1116" height="850" alt="Image" src="https://github.com/user-attachments/assets/1c3415a5-945f-415c-8f88-317848b54e7d" />

[Playground](https://www.typescriptlang.org/play/?isolatedDeclarations=true#code/JYOwLgpgTgZghgYwgAgDLAM5gJKQLbIDeAUMsiHHhAFzJZSgDmA3KcgCYQYIMAOYwAPYha9JqwC+xYhAAevQVDDIEwrMhiDByALxE2cWiTJkA9KeRgoAT0vaw13igAGFKs+QALaCgAUAd2AwT0EAV2VnZwBKajYpCWQ4DBU1MFZiTW0MOAEMGGAuZAAlCFUodgAeMRBGABo0TBx8AD5mIA)

### 🙂 Expected behavior

Either the original code would work without errors.

Or the suggested workaround would work with intellisense.

### Additional information about the issue

Sometimes `isolatedDeclarations` is mentioned with tsgo or improving perf for monorepos. So, it'll be great if it can be made easier for developers.

Thanks a lot.

**Comment by Andarist** [comment]
@dragomirtitian has shown a good example [here](https://github.com/microsoft/TypeScript/issues/58397#issuecomment-2090895054) why `satisfies` might change the expression's type... but I wonder, in that case the function already contains an error because it's not fully-annotated. I can't think of any other example that would be affected by `satisfies`' contextual typing that wouldn't involve functions.

**Comment by Meligy** [comment]
The key problem is that there's no way to use `as const` and `satisfies` together, and dare I say that this is a common use?

I'm not hung on to either of the 2 options (making the original code work, or making the alternative code work).

Just need "some way" to enable the pattern. I maintain code that many developers contribute to, and trying to introduce `isolatedDeclarations` feels like I'm just giving my developrs pain if simple patterns like this one start to fail.

A 3rd option might be making `satsifies` before `as const` works.
But not saying that this has to be the answer also. Maybe it can be blocked by the same issue or something else.
If you can come up with a 4th option that makes the pattern work, that'll be equally cool.

All I'm saying is that the 2nd option is also so broken in practice.

Although, now that you mention it, I wonder if `satisfies` should change the type in this case at all.
Maybe that is the problem.
But if the team feels strongly against that, it's not a hill I am willing to die on also.

**Comment by Andarist** [comment]
> Although, now that you mention it, I wonder if satisfies should change the type at all.

Regardless of any opinions on this, it's a ship that already has sailed and it was a conscious design decision at the time this feature was introduced ;p

> A 3rd option might be making satsifies before as const works.

Tbf, that wouldn't solve your problem. Even if this would be allowed, I'd just assume the same outcomes (those 2 to be commutative).

> I maintain code that many developers contribute to, and trying to introduce isolatedDeclarations feels like I'm just giving my developrs pain if simple patterns like this one start to fail.

My personal opinion is that ID is just not worth it in practice - but I also can't say I have worked on any project that would push me to consider adopting it. I've seen TS codebases suffering from perf problems but usually I'd work around them using other approaches. The rewrite in Go should also push the ceiling much further and I don't expect many codebases to adopt ID in practice. I'm not denying that in some massive massive monorepos this can make a difference that would be worth considering

**Comment by Meligy** [comment]
Actually what first introduced me to ID was that tsgo seemed to act like it's on, always.

So, I was looking into implementing it as prep for tsgo, and maybe even we'd use tsgo with something like rolldown / oxlint, etc.

Otherwise, I'd have agreed yeah.

Although to be honest, tsgo does not act 100% the same way as ID:

1. It requires some exported functions (react components especially) to have an explicit return type like ID, even if ID is false
2. But it does not break this example with ID false, according to my anecdotal evidence

But tsgo is a different story I guess.

> > Although, now that you mention it, I wonder if satisfies should change the type at all.
>
> Regardless of any opinions on this, it's a ship that already has sailed and it was a conscious design decision at the time this feature was introduced ;p

Well, how far is TS 6.0? 😉

As always, not saying that this has to be the answer.

> > A 3rd option might be making satsifies before as const works.
>
> Tbf, that wouldn't solve your problem. Even if this would be allowed, I'd just assume the same outcomes (those 2 to be commutative).

What I was thinking is if the compiler can recognise the order, which was not allowed before, it can have other rules when processing it (update: but again, not saying that this is "THE answer").

If the answer is: ID is bad anyway, then, not great, but I'll have an issue to reference if it comes up in conversations 😐

Although it might be harder to implement prepping a codebase for tsgo without switching to it yet, without ID.

**Comment by Andarist** [comment]
> It requires some exported functions (react components especially) to have an explicit return type like ID, even if ID is false

This feels like a bug in TS Go. Remember, it's a work in progress. There are some intended functional differences between Strada (TS-in-TS) and Corsa (TS-in-Go) but the rule of thumb is that Corsa should behave the same way as Strada does

> Well, how far is TS 6.0? 😉

I don't think this is a breaking change TS team would ever be willing to make. Versioning here doesn't follow semver the same way as other projects follow those guidelines.

> If the answer is: ID is bad anyway, then, not great, but I'll have an issue to reference if it comes up in conversations 😐

I have not said ID is bad per-se. I just don't see it being worth it in the projects I'm working on. This is not an official statement, I'm not a TS team member 😉

**Comment by Meligy** [comment]
Well. Not sure discussing the value of ID is valuable in itself if not in the context of saying that this issue is not important because ID is not something worth using anyway.

**Comment by RyanCavanaugh** [comment]
I don't think anyone would claim that ID doesn't have its drawbacks. It's intentionally narrowly designed to enforce a very particular and very strict invariant. If you don't *need* it for its performance gains, there's literally no reason to turn it on (from the 5.5 release blog):

> We also feel it is worth calling out that isolatedDeclarations should be adopted on a case-by-case basis. There are some developer ergonomics that are lost when using isolatedDeclarations, and thus it may not be the right choice if your setup is not leveraging the two scenarios mentioned earlier

A lot of the hesitation in adding this feature *was* that it has a fairly high DX penalty, and this is part of that cost. There was never any intention of making it so that every use case always had an ID-compatible way to do the exact same thing -- this isn't really a possible invariant, and there's no way we're going to add new syntactic forms if the only endgoal is slightly changing which of those scenarios can be met. ID is a pay-for-play feature, otherwise it'd be on by default.

For examples like this one, the "some way" is to make a temporary edit to your code to get the right Intellisense. I don't see any path forward here that doesn't "break" a fundamental promise that ID, `satisfies`, or `const` provides. Alternatively, you could imagine the LS could recognize separated `satisfies`/`const` assertions as special, but that kind of functionality seems wildly complicated and unlikely to be worth the maintenance cost.
